### PR TITLE
[SPARK-44760][INFRA] Fix list index out of range for JIRA resolution in merge_spark_pr

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This PR fixes list index out-of-range error for the merge_spark_pr script

The error occurs when the branch we merge into does not have a jira project version.

```python
>>> versions = sorted(versions, key=lambda x: x.name, reverse=True)
>>> versions
[<JIRA Version: name='4.0.0', id='12353359'>, <JIRA Version: name='3.5.1', id='12353495'>, <JIRA Version: name='3.5.0', id='12352848'>, <JIRA Version: name='3.4.2', id='12353368'>, <JIRA Version: name='3.3.4', id='12353505'>, <JIRA Version: name='3.3.3', id='12352932'>]
>>> list(filter(lambda x: x.name.startswith(""), versions))[-1]
<JIRA Version: name='3.3.3', id='12352932'>
>>> list(filter(lambda x: x.name.startswith("3.2"), versions))[-1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```

Unlike other archived branches or versions, v3.2.5 is missing from our JIRA. This crushes the merging scripts during JIRA resolution if branch-3.2 is chosen.

### Why are the changes needed?

happy merging

### Does this PR introduce _any_ user-facing change?

no, infra only change

### How was this patch tested?

manually tested 

```python
>>> default_fix_versions = []
>>> for b in merge_branches:
...     if b == "master":
...         default_fix_versions.append(versions[0])
...     else:
...         found = False
...         for v in versions:
...             if v.name.startswith(b.replace("branch-", "")):
...                 default_fix_versions.append(v)
...                 found = True
...         if not found:
...             print(
...                 "Target version for %s is not found on JIRA, it may be archived or "
...                 "not created. Skipping it." % b)
...
Target version for branch-3.2 is not found on JIRA, it may be archived or not created. Skipping it.
>>> default_fix_versions
[<JIRA Version: name='4.0.0', id='12353359'>, <JIRA Version: name='3.5.0', id='12352848'>, <JIRA Version: name='3.5.1', id='12353495'>, <JIRA Version: name='3.4.2', id='12353368'>, <JIRA Version: name='3.3.3', id='12352932'>, <JIRA Version: name='3.3.4', id='12353505'>]
>>> quit
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @dongjoon-hyun @cloud-fan @HyukjinKwon, thanks

**Review by cloud-fan — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
ah, need to fix the lint by `dev/reformat-python`

**Comment by yaooqinn** [comment]
Thank you for the reminder, @zhengruifeng.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn .

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you all.

File: dev/merge_spark_pr.py [hunk 1]
```diff
@@ -280,14 +271,30 @@ def resolve_jira_issue(merge_branches, comment, default_jira_id=""):
...
+    ]
     versions = sorted(versions, key=lambda x: x.name, reverse=True)
-    versions = list(filter(lambda x: x.raw["released"] is False, versions))
-    # Consider only x.y.z versions
-    versions = list(filter(lambda x: re.match(r"\d+\.\d+\.\d+", x.name), versions))
 
-    default_fix_versions = list(
-        map(lambda x: fix_version_from_branch(x, versions).name, merge_branches)
-    )
+    default_fix_versions = []
```
Comment (HyukjinKwon) [inline dev/merge_spark_pr.py:282]: Actually seems this breaks the script:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/jira/resources.py", line 222, in __getattr__
    return self[item]  # type: ignore
TypeError: 'Version' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.../spark/./dev/merge_spark_pr.py", line 606, in <module>
    main()
  File "/.../spark/./dev/merge_spark_pr.py", line 590, in main
    resolve_jira_issues(title, merged_refs, jira_comment)
  File "/.../spark/./dev/merge_spark_pr.py", line 397, in resolve_jira_issues
    resolve_jira_issue(merge_branches, comment, jira_id)
  File "/.../spark/./dev/merge_spark_pr.py", line 303, in resolve_jira_issue
    (major, minor, patch) = v.split(".")
  File "/usr/local/lib/python3.9/site-packages/jira/resources.py", line 227, in __getattr__
    raise AttributeError(
AttributeError: <class 'jira.resources.Version'> object has no attribute 'split' ('Version' object is not subscriptable)
```

The reason seems that `default_fix_versions` has to be `Version` instance, and you should get the version from it via `.name`.

I am reverting this for now.

Comment (cloud-fan) [reply dev/merge_spark_pr.py:282]: Next time please use the updated merge script to merge the merge script update PR...

Comment (dongjoon-hyun) [reply dev/merge_spark_pr.py:282]: Oh, ya, it's my bad. Sorry for the trouble, @HyukjinKwon and @cloud-fan .

Comment (yaooqinn) [reply dev/merge_spark_pr.py:282]: This should be my fault, I didn't validate the code below sufficiently.

**Comment by yaooqinn** [comment]
Merged https://github.com/apache/spark/pull/42407, https://github.com/apache/spark/pull/42404, https://github.com/apache/spark/pull/42442 by last updated script. It shall be fine now

**Review by HyukjinKwon — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master, thanks @zhengruifeng @dongjoon-hyun @cloud-fan @HyukjinKwon

**Pull request closed without merging**